src/StableBuffer.mo

```diff
@@ -86,88 +89,95 @@ module {
 
   /// Resets the buffer.
-  public func clear<X>(buffer: StableBuffer<X>): () {
+  public func clear<X>(buffer : StableBuffer<X>) : () {
     buffer.count := 0;
   };
 
   /// Returns a copy of this buffer.
-  public func clone<X>(buffer: StableBuffer<X>) : StableBuffer<X> {
+  public func clone<X>(buffer : StableBuffer<X>) : StableBuffer<X> {
     let c = initPresized<X>(buffer.elems.size());
-    var i = 0;
+    var i = vals(buffer);
+
     label l loop {
-      if (i >= buffer.count) break l;
-      add(c, buffer.elems[i]);
-      i += 1;
+      switch (i.next()) {
+        case null break l;
+        case (?x) { add(c, x) };
+      };
     };
-    c
+
+    c;
   };
 
   /// Returns an `Iter` over the elements of this buffer.
-  public func vals<X>(buffer: StableBuffer<X>) : { next : () -> ?X } = object {
+  public func vals<X>(buffer : StableBuffer<X>) : { next : () -> ?X } = object {
     var pos = 0;
     public func next() : ?X {
       if (pos == buffer.count) { null } else {
-        let elem = ?buffer.elems[pos];
+        let elem = buffer.elems[pos];
         pos += 1;
-        elem
-      }
-    }
+        elem;
+      };
+    };
   };
 
   /// Creates a Buffer from an Array
-  public func fromArray<X>(xs: [X]): StableBuffer<X> {
-    let ys: StableBuffer<X> = initPresized(xs.size());
+  public func fromArray<X>(xs : [X]) : StableBuffer<X> {
+    let ys : StableBuffer<X> = initPresized(xs.size());
     for (x in xs.vals()) {
       add(ys, x);
     };
 
-    ys
+    ys;
   };
 
   /// Creates a new array containing this buffer's elements.
-  public func toArray<X>(buffer: StableBuffer<X>) : [X] =
-    // immutable clone of array
-    Prim.Array_tabulate<X>(
-      buffer.count,
-      func(x : Nat) : X { buffer.elems[x] }
-    );
+  public func toArray<X>(buffer : StableBuffer<X>) : [X] =
+  // immutable clone of array
+  Prim.Array_tabulate<X>(
+    buffer.count,
+    func(x : Nat) : X {
+      get(buffer, x);
+    },
+  );
 
   /// Creates a mutable array containing this buffer's elements.
-  public func toVarArray<X>(buffer: StableBuffer<X>) : [var X] {
+  public func toVarArray<X>(buffer : StableBuffer<X>) : [var X] {
     if (buffer.count == 0) { [var] } else {
-      let a = Prim.Array_init<X>(buffer.count, buffer.elems[0]);
+      let a = Prim.Array_init<X>(buffer.count, get(buffer, 0));
       var i = 0;
       label l loop {
         if (i >= buffer.count) break l;
-        a[i] := buffer.elems[i];
+        a[i] := get(buffer, i);
         i += 1;
       };
-      a
-    }
+
+      a;
+    };
   };
 
   /// Gets the `i`-th element of this buffer. Traps if  `i >= count`. Indexing is zero-based.
-  public func get<X>(buffer: StableBuffer<X>, i : Nat) : X {
-    assert(i < buffer.count);
-    buffer.elems[i]
+  public func get<X>(buffer : StableBuffer<X>, i : Nat) : X {
+    switch (buffer.elems[i]) {
+      case (?elem) elem;
+      case (null) Prim.trap("Buffer index out of bounds in get");
```

Comment:
One of the issues with moving this to `optional` type land is that now a developer can insert a few items, then set index 0 to `null`, and then this would trap on `buffer.get(0)`. I guess you can already mess around with the mutable `count` field now though 🤔 

The "new" implementation in base is a class based implementation (requires transformation in order to achieve stability), but has the benefit of the elements and count being private instance variables that can't be modified by the developer.

I'm sort of on the fence about just ripping all the new Buffer implementation out of the class and into functions (just like I did for the old Buffer implementation), but that new module is just so huge 😆 

Another thing to consider is that this would be a breaking change in terms of the underlying data structure, so I'll need to bump the major version (not a huge deal).

Reply:
I hadn't thought of that 😲. It's hard to tamper-proof the internal state. I think the alternative would be to switch to a functional implementation that returns a new state after every function.

I was wondering if this same issue could happen if the value set by the user is also an optional type, making the internal type a nested option (`??X`). But after testing this in the [Motoko playground](https://m7sm4-2iaaa-aaaab-qabra-cai.raw.ic0.app/?tag=2296341354) it works quite well.

Yes, I agree that the functions included in the `Buffer` module are a lot and don't need to be added to this one. I would only like to add the essential function in the `Buffer` class like `insert` and `remove`. 

I don't have a lot of experience publishing libs for users. So I don't have an developed sense of what's considered a breaking or a minor change. Why is it a breaking change in this case? When to the user, none of the functions really change.

Reply:
> "I don't have a lot of experience publishing libs for users. So I don't have a developed sense of what's considered a breaking or a minor change. Why is it a breaking change in this case? When to the user, none of the functions really change."

So even though the interfaces are the same, the underlying types are different.

For example, the [current Buffer class](https://github.com/dfinity/motoko-base/blob/aafcdee0c8328087aeed506e64aa2ff4ed329b47/src/Buffer.mo#L67) is a class and therefore is **not** stable (requires preupgrade/postupgrade).

Also, the class-based implementation internals (elems, count) are private, so they are not directly accessible by developers. If I'm using a class-based implementation, I know that by changing the types, there's no way that end users are currently directly pulling from and using the data structure internals in their code, but are going through the class-based interfaces.

If I make this change from [var X] to [var ?X], then this means developers who have an existing buffer

`stable let myBuffer: StableBuffer<X> = ...`

which has underlying type

```
public type StableBuffer<X> = {
  initCapacity : Nat;
  var count : Nat;
  var elems : [var ?X];
};
```

Will now become underlying type

```
public type StableBuffer<X> = {
  initCapacity : Nat;
  var count : Nat;
  var elems : [var ?X];
};
```

So you'll end up receiving an error when you attempt to re-deploy. Something like:

```
WARNING!
Stable interface compatibility check failed for canister 'example'.
Upgrade will either FAIL or LOSE some stable variable data.

(unknown location): Compatibility error [M0170], stable variable myBuffer of previous type
  {var count : Nat; var elems : [var Nat]; initCapacity : Nat}
cannot be consumed at new type
  {var count : Nat; var elems : [var ?Nat]; initCapacity : Nat}

Do you want to proceed? yes/No
```

Check out the [migration testing I did](https://github.com/canscale/StableBuffer/commit/f4a8c13aa263687aebe691565016c426048d5582), following these steps.

1. Navigate to the `example` folder and `dfx deploy` the example.
2. Enter a few values into the buffer via `addNatToBuffer()` and test out that they were inserted
3. Sub in the new stable buffer
  a. Comment out the old version (first half of `src/StableBuffer.mo` file Lines 1-174), and uncomment the new StableBuffer.mo implementation (second half of src/StableBuffer.mo Lines 180-end)
4. Try to redeploy - you'll notice the breaking change in the example/src/main.mo actor under the `printBuffer()` function.
5. Sub in the new version of the actor (comment out `example/src/main.mo` actor at first half of file, uncomment second actor at bottom of file)
6. Try to re-deploy. You'll get the interface compatibility warning I showed above ☝️. This will scare away many developers. 
7. Continue through with the deployment and type yes. You'll notice that despite the warning, the migration was successful (this was baffling to me, but I think it has something to do with the `postUpgrade()` deserialization)

Despite the fact that this upgrade was stable, it still is a breaking change - it broke an API (required an update to the logic) due to the fact that the buffer internals are accessible.

An idea came to me while reviewing this PR. Check out [this](https://github.com/canscale/StableBuffer/commit/f4a8c13aa263687aebe691565016c426048d5582#diff-38735e3d38cb2f65640007a8695265c9d43a5ff4dd66a1dbf2aa9f64cd6272b2). What do you think of just adding a new module called `StableOptBuffer` for your use case? This way no one gets broken, and they can migrate if they choose to rather than being forced to by the library overwriting the previous StableBuffer module.